### ISSUE TITLE
NET-1998 Code Coverage: exclusions in .runsettings file are ignored

### ISSUE THREAD

**Issue by berndku** [SonarSource/sonar-dotnet#1487]
### Description

after upgrading SonarQube Server from 5.6 to 6.7.4.38452 and C# plugin to 7.2 (build 5463), Code Coverage is still imported from VSTS build, but provided exclusions in .runsettings file for VSTest are ignored. With SonarQube 5.6 the .runsettings file exclusions were correctly applied.

### Repro steps
VSTS SonarQube Build tasks used. 
VSTest task has Code Coverage enabled and a .runsettings file specified in the property "Settings file". 

### Expected behavior
Code Coverage in SonarQube should recognize exlusions from VSTest .runsettings file. 

### Actual behavior

Code Coverage is imported into SonarQube, but exclusions are not recognized.

### Related information

* SonarC# Version : 7.2
* Visual Studio Version : 15.7.3
* Scanner for MSBuild: VSTS Tasks Version 3

**Comment by rliberoff** [comment]
We are having exactly the same issue.

Please fix it!!!!!!

**Comment by millusdk** [comment]
Are there any updates on this? We are experiencing the same issue in 6.7

**Comment by Evangelink** [comment]
Hi there,

I finally managed to spend some time looking at this issue.

Starting with the version 6.6 of SonarC#, we are now processing a new metric called "executable lines of code" which is used in addition to the coverage info to track lines covered/uncovered. As we don't process (and we never did) the `.runsettings` file, all files excluded in this configuration won't have any coverage info BUT using the executable lines of code will appear as un-covered on SonarQube.

As a short-term answer, you will need to provide a new parameter (`sonar.coverage.exclusion=LIST_OF_FILES`) to the scanner for msbuild (or the prepare task of TFS/VSTS) OR configure your project on the SonarQube/SonarCloud project administration page to exclude all files you have excluded in your `.runsettings` file. Note that you can only exclude a file or a group of files BUT not something with a lower granularity.

As a mid-term/long-term solution, we will investigate whether we should automatically exclude all files excluded in the `.runsettings` file.

**Comment by puddlewitt** [comment]
Closed?

**Comment by andrei-epure-sonarsource** [comment]
Before moving forward with this, we'll have to check the impact of recently handling of `.gitignore`, and then come back with a plan

**Comment by JFMG** [comment]
@andrei-epure-sonarsource How far is the plan?

**Comment by StingyJack** [comment]
@andrei-epure-sonarsource  - Would this apply to SonarCloud as well?

**Comment by andrei-epure-sonarsource** [comment]
I'm coming back on this topic after a long pause (sorry for that!).

@berndku @rliberoff @millusdk @puddlewitt @JFMG @StingyJack For clarification of your usecase(s): are you looking for excluding projects / files just for code coverage or from analysis in general? In other words:
- would your expectation be for us to read the runsettings.xml file and exclude those files completely from analysis (like the behavior @Evangelink mentioned [here](https://github.com/SonarSource/sonar-dotnet/issues/1487#issuecomment-437404878) and in [the docs](https://docs.sonarqube.org/latest/project-administration/narrowing-the-focus/), but out of the box based on what's excluded in the runsettings file)
- **or** would you expect these files to be analyzed by our analyzers, but excluded only from code coverage?

Thank you 🙏 

<hr>

> Would this apply to SonarCloud as well?

@StingyJack - if this were implemented, yes, it would apply for both SonarQube and SonarCloud.

**Comment by berndku** [comment]
Analyzed, but excluded from code coverage.

**Comment by StingyJack** [comment]
Just excluded from code coverage analysis, which is what that runsettings file declares for the VS enterprise code coverage tooling.

**Comment by JFMG** [comment]
I agree with the previous comments. The runsettings file should just do what it says in the official documentation https://docs.microsoft.com/en-US/visualstudio/test/configure-unit-tests-by-using-a-dot-runsettings-file That means just exclude everything what is described in the runsettings file from the code coverage.

**Comment by StingyJack** [comment]
At the bottom of that docs page @JFMG shared is a link to [this page](https://docs.microsoft.com/en-US/visualstudio/test/customizing-code-coverage-analysis?view=vs-2019#include-or-exclude-assemblies-and-members) that details the types that can be included or excluded for coverage analysis. 

In the past I've needed three of those filters. 
**ModulePath** - to either whitelist the assemblies to check or to exclude unit tests from being analyzed for coverage
**Attribute** - to filter out the ones mentioned in the [example runsettings](https://docs.microsoft.com/en-US/visualstudio/test/customizing-code-coverage-analysis?view=vs-2019#sample-runsettings-file) 
**Source** - to exclude code gen files by extension (`*.g.cs`)

**Comment by andrei-epure-sonarsource** [comment]
Thanks for the replies, @berndku , @JFMG , @StingyJack.

I'm thinking out loud: if we were to implement a partial solution (we like to deliver value incrementally) , for example starting with the **Source** filter only:
 - would you find it more confusing that the **Source** filter is taken into consideration, and the **ModulePath** and **Attribute** (and others) are not?
 - would you check the build logs in case the behavior from SonarQube/SonarCloud is different than what you'd expect (we'd mention in the build logs that only filter X is supported, and the rest are ignored)?

Another question: in my understanding, there would be only one `.runsettings` file per analysis, right?

**Comment by mack0196** [comment]
What is the status of respecting the runsettings in sonar analysis?

**Comment by andrei-epure-sonarsource** [comment]
Hi @mack0196 - this ticket is still open. I've asked a few questions here which I didn't get reply to. 

We don't have this on our near-future roadmap.

(cc @tom-howlett-sonarsource )

**Comment by rchaidez-cricut** [comment]
Please respect what runsettings file has for exclusions

**Comment by mack0196** [comment]
My team and I are still interested if you could have sonar respect the exclusions in our runsettings file.

**Comment by wilkovanderveen** [comment]
Any other options than waiting for this change?

**Comment by nathfy** [comment]
@andrei-epure-sonarsource Any update on when this will be picked up by sonar? At the moment we need to add exclusion to both `CodeCoverage.runsettings` and in sonar cloud > General settings > B. Code Coverage Exclusions

**Comment by meyerjg** [comment]
Having the same issue/behavior, would like to see an update for this

**Comment by almolo1** [comment]
This is still relevant in 2024, .runsettings file should be respected

**Comment by pavel-mikula-sonarsource** [comment]
cc @denis-troller to raise awareness

**Comment by williamoconnorme** [comment]
This is the top voted issue on this tracker and is nearing it's 7th birthday. Is there any timeline on when or if this will be prioritised? The alternatives of setting `sonar.coverage.exclusions` is lacking in what can be configured with the limited regex matching patterns as you can't easily include a module path like what can be set in runsettings file. This could be mitigated with a property such as `sonar.coverage.inclusions` but this does not exist today.

**Comment by pavel-mikula-sonarsource** [comment]
Moved to [NET-1998](https://sonarsource.atlassian.net/browse/NET-1998)

[NET-1998]: https://sonarsource.atlassian.net/browse/NET-1998?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
